### PR TITLE
Support binding OAuth2 user automatically

### PR DESCRIPTION
**Pull request by JohnNiang**
#### What type of PR is this?

/kind improvement
/area core
/milestone 2.20.x

#### What this PR does / why we need it:

This PR add support for binding OAuth2 user automatically. So we can remove the user-binding page.

Please note that those changes may break the OAuth2 and SocialLogin plugins.

#### Special notes for your reviewer:

Build OAuth2 plugin from <https://github.com/halo-sigs/plugin-oauth2/pull/64> or use [plugin-oauth2-1.0.4-SNAPSHOT.zip](https://github.com/user-attachments/files/17177592/plugin-oauth2-1.0.4-SNAPSHOT.zip) I built.

- Bind after logging in

1. Log in Halo with username and password method
2. Try to unbind OAuth2 user
3. Bind OAuth2 user again

- Initially bind without logging in

1. Go to login page
2. Log in with OAuth2 method and you will be redirected to login page
3. Log in with username and password method
4. See the result of binding

- Log in with OAuth2 method after binding

1. Go  to login page
2. Log in with OAuth2 method and you will be redirected to uc page directly

#### Does this PR introduce a user-facing change?

```release-note
支持自动绑定 OAuth2 登录用户
```

### CODE REVIEW

File: application/src/main/java/run/halo/app/core/user/service/impl/UserConnectionServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    private Mono<UserConnection> updateUserConnection(UserConnection connection,
+        OAuth2User oauth2User) {
+        connection.getSpec().setUpdatedAt(clock.instant());
+        updateUserInfo(connection.getMetadata(), oauth2User);
+        return client.update(connection);
+    }
+
+    @Override
+    public Mono<UserConnection> getAndUpdateUserConnection(String registrationId,
```
Comment (guqing) [inline application/src/main/java/run/halo/app/core/user/service/impl/UserConnectionServiceImpl.java:65]: 感觉这个名称有歧义,会理解为先查询得到的数据返回后再更新,是否可以直接改为 updateUserConnection,毕竟这里的逻辑就是根据 OAuth2User 去更新 UserConnection

File: application/src/main/java/run/halo/app/core/user/service/impl/UserConnectionServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    public UserConnectionServiceImpl(ReactiveExtensionClient client) {
+        this.client = client;
+    }
+
+    void setClock(Clock clock) {
+        this.clock = clock;
+    }
+
+    @Override
+    public Mono<UserConnection> createUserConnection(
```
Comment (guqing) [inline application/src/main/java/run/halo/app/core/user/service/impl/UserConnectionServiceImpl.java:38]: 这里需要考虑一个用户只能绑定同一平台的一个账号的问题,参考 [https://github.com/halo-sigs/plugin-oauth2/pull/63 和 ](https://github.com/halo-sigs/plugin-oauth2/pull/59),比如 zhangsan 有 githubA 和 githubB 两个账号,如果已经绑定了其中之一不能再绑定另外一个

Comment (JohnNiang) [reply application/src/main/java/run/halo/app/core/user/service/impl/UserConnectionServiceImpl.java:39]: Done.

**Review by ruibaby — COMMENTED** [review]
已测试登录和注册之后都可以正常绑定。

/lgtm

**Review by guqing — APPROVED** [review]
/lgtm

**Pull request merged**